tools/tdbg/factory.go

```diff
@@ -0,0 +1,216 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tdbg
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/urfave/cli/v2"
+	"go.temporal.io/server/api/adminservice/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
+
+	"go.temporal.io/server/common/auth"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+)
+
+var netClient HttpGetter = &http.Client{
+	Timeout: time.Second * 10,
+}
+
+// HttpGetter defines http.Client.Get(...) as an interface so we can mock it
+type HttpGetter interface {
+	Get(url string) (resp *http.Response, err error)
+}
+
+// ClientFactory is used to construct rpc clients
+type ClientFactory interface {
+	AdminClient(c *cli.Context) adminservice.AdminServiceClient
+	HealthClient(c *cli.Context) healthpb.HealthClient
+}
+
+type clientFactory struct {
+	logger log.Logger
+}
+
+// NewClientFactory creates a new ClientFactory
+func NewClientFactory() ClientFactory {
+	logger := log.NewCLILogger()
+
+	return &clientFactory{
+		logger: logger,
+	}
+}
+
+// AdminClient builds an admin client.
+func (b *clientFactory) AdminClient(c *cli.Context) adminservice.AdminServiceClient {
+	connection, _ := b.createGRPCConnection(c)
+
+	return adminservice.NewAdminServiceClient(connection)
+}
+
+// HealthClient builds a health client.
+func (b *clientFactory) HealthClient(c *cli.Context) healthpb.HealthClient {
+	connection, _ := b.createGRPCConnection(c)
+
+	return healthpb.NewHealthClient(connection)
+}
+
+func (b *clientFactory) createGRPCConnection(c *cli.Context) (*grpc.ClientConn, error) {
+	hostPort := c.String(FlagAddress)
+	if hostPort == "" {
+		hostPort = localHostPort
+	}
+
+	tlsConfig, err := b.createTLSConfig(c)
+	if err != nil {
+		return nil, err
+	}
+
+	grpcSecurityOptions := grpc.WithTransportCredentials(insecure.NewCredentials())
+
+	if tlsConfig != nil {
+		grpcSecurityOptions = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
+	}
+
+	dialOpts := []grpc.DialOption{
+		grpcSecurityOptions,
+	}
+
+	connection, err := grpc.Dial(hostPort, dialOpts...)
+	if err != nil {
+		b.logger.Fatal("Failed to create connection", tag.Error(err))
+		return nil, err
+	}
+	return connection, nil
+}
+
+func (b *clientFactory) createTLSConfig(c *cli.Context) (*tls.Config, error) {
```

Comment:
@yiminc, do you think we need TLS for tdbg at all?

Reply:
Good question. TLS probably is not what tdbg would like to teal with. However, I don't think we can skip TLS if it is enabled by server.